File: src/state.rs
use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AmuxState {
    pub version: u32,
    pub selected: usize,
    pub view_mode: String,
    pub windows: Vec<WindowEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WindowEntry {
    pub position: usize,
    pub tmux_window: String,
    pub label: String,
}

impl AmuxState {
    /// Create a new default state.
    pub fn new() -> Self {
        AmuxState {
            version: 1,
            selected: 0,
            view_mode: "grid".to_string(),
            windows: Vec::new(),
        }
    }

    /// Load state from `path`. Returns `None` if the file is missing or corrupt.
    /// Currently unused in v2, but kept for future state persistence features.
    #[allow(dead_code)]
    pub fn load(path: &Path) -> Option<Self> {
        let contents = fs::read_to_string(path).ok()?;
        serde_json::from_str(&contents).ok()
    }

    /// Save state to `path` atomically via a temporary file + rename.
    /// Parent directories are created if they don't exist.
    pub fn save(&self, path: &Path) -> Result<()> {
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        // Write to a sibling temp file, then rename for atomicity.
        let tmp_path = path.with_extension("json.tmp");
        let json = serde_json::to_string_pretty(self)?;
        fs::write(&tmp_path, &json)?;
        fs::rename(&tmp_path, path)?;
        Ok(())
    }

    /// Return the default state file path: `~/.amux/state.json`.
    pub fn default_path() -> PathBuf {
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("."));
        home.join(".amux").join("state.json")
    }

    /// Reconcile saved state against the current live tmux windows.
    ///
    /// - Saved windows that still exist in `live_windows` are kept in their
    ///   saved position order.
    /// - Live windows not present in the saved state are appended at the end.
    /// - `selected` is clamped to the valid range.
    ///
    /// Currently unused in v2, but kept for future state persistence features.
    #[allow(dead_code)]
    pub fn reconcile(&self, live_windows: &[String]) -> AmuxState {
        // Keep saved windows that still appear in the live set, preserving order.
        let mut result_windows: Vec<WindowEntry> = self
            .windows
            .iter()
            .filter(|w| live_windows.contains(&w.tmux_window))
            .cloned()
            .collect();

        // Collect the set of tmux_window names already in result.
        let already_saved: std::collections::HashSet<String> = result_windows
            .iter()
            .map(|w| w.tmux_window.clone())
            .collect();

        // Collect live windows not yet represented.
        let next_position = result_windows.len();
        let new_entries: Vec<WindowEntry> = live_windows
            .iter()
            .filter(|w| !already_saved.contains(*w))
            .enumerate()
            .map(|(i, live)| WindowEntry {
                position: next_position + i,
                tmux_window: live.clone(),
                label: live.clone(),
            })
            .collect();

        // Append the new entries.
        result_windows.extend(new_entries);

        // Reindex positions to be contiguous.
        for (i, w) in result_windows.iter_mut().enumerate() {
            w.position = i;
        }

        // Clamp selected index.
        let max_idx = result_windows.len().saturating_sub(1);
        let selected = self.selected.min(max_idx);

        AmuxState {
            version: self.version,
            selected,
            view_mode: self.view_mode.clone(),
            windows: result_windows,
        }
    }
}

impl Default for AmuxState {
    fn default() -> Self {
        Self::new()
    }
}

File: app/Sources/AmuxLib/State.swift
// State.swift -- Persistent state for amux sessions.
//
// Saves and loads session state (selected pane, view mode, window list)
// to ~/.amux/state.json using atomic write (temp file + rename).

import Foundation

public struct WindowEntry: Codable {
    public var position: Int
    public var tmuxWindow: String
    public var label: String

    public init(position: Int, tmuxWindow: String, label: String) {
        self.position = position
        self.tmuxWindow = tmuxWindow
        self.label = label
    }

    enum CodingKeys: String, CodingKey {
        case position
        case tmuxWindow = "tmux_window"
        case label
    }
}

public struct AmuxState: Codable {
    public var version: Int
    public var selected: Int
    public var viewMode: String
    public var windows: [WindowEntry]

    enum CodingKeys: String, CodingKey {
        case version
        case selected
        case viewMode = "view_mode"
        case windows
    }

    /// Create a new default state.
    public init() {
        version = 1
        selected = 0
        viewMode = "grid"
        windows = []
    }

    /// Load state from a file path. Returns nil if the file is missing or corrupt.
    public static func load(from path: URL) -> AmuxState? {
        guard let data = try? Data(contentsOf: path) else { return nil }
        return try? JSONDecoder().decode(AmuxState.self, from: data)
    }

    /// Save state to a file path atomically via a temporary file + rename.
    /// Parent directories are created if they don't exist.
    public func save(to path: URL) throws {
        if let parent = path.deletingLastPathComponent() as URL? {
            try FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true)
        }

        // Write to a sibling temp file, then rename for atomicity.
        let tmpPath = path.deletingPathExtension().appendingPathExtension("json.tmp")
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(self)
        try data.write(to: tmpPath)
        _ = try FileManager.default.replaceItemAt(path, withItemAt: tmpPath)
    }

    /// Return the default state file path: ~/.amux/state.json
    public static var defaultPath: URL {
        URL(fileURLWithPath: NSHomeDirectory())
            .appendingPathComponent(".amux")
            .appendingPathComponent("state.json")
    }

    /// Reconcile saved state against the current live tmux windows.
    ///
    /// - Saved windows that still exist in liveWindows are kept in their saved order.
    /// - Live windows not present in the saved state are appended at the end.
    /// - selected is clamped to the valid range.
    public func reconcile(liveWindows: [String]) -> AmuxState {
        // Keep saved windows that still appear in the live set, preserving order.
        var resultWindows = windows.filter { liveWindows.contains($0.tmuxWindow) }

        // Collect the set of tmux_window names already in result.
        let alreadySaved = Set(resultWindows.map(\.tmuxWindow))

        // Collect live windows not yet represented.
        let nextPosition = resultWindows.count
        let newEntries = liveWindows
            .filter { !alreadySaved.contains($0) }
            .enumerated()
            .map { (i, live) in
                WindowEntry(position: nextPosition + i, tmuxWindow: live, label: live)
            }

        resultWindows.append(contentsOf: newEntries)

        // Reindex positions to be contiguous.
        for i in resultWindows.indices {
            resultWindows[i].position = i
        }

        // Clamp selected index.
        let maxIdx = max(resultWindows.count - 1, 0)
        let clampedSelected = min(selected, maxIdx)

        var state = AmuxState()
        state.version = version
        state.selected = clampedSelected
        state.viewMode = viewMode
        state.windows = resultWindows
        return state
    }
}
